jax/core.py

```diff
@@ -1579,6 +1582,26 @@ def axis_frame(axis_name):
     else:
       raise NameError("unbound axis name: {}".format(axis_name))
 
+  @contextmanager
+  def extend_axis_env_poly(axis_name, size: int):
```

Comment:
Can we just use the same `trace_state.axis_env` and `extend_axis_env` function?

Reply:
I think it's a little bit tricky to use the same `axis_env`. There are two options if you're going to do that:

1. Allow an AxisEnvFrame to contain _both_ a padded (hard) axis size and a polymorphic one. So for a polymorphic axis you'd have something like
    ```python
    trace_state.axis_env = [('i', 9, b ** 2 + a + 1)]
    ```
2. Have two AxisEnvFrames for polymorphic axes, one containing the hard axis size and the other containing the polymophic one, i.e.
    ```python
    trace_state.axis_env = [('i', 9),
                            ('i', b ** 2 + a + 1)]
    ```
Option 2 is a bit ugly IMO, option 1 is do-able but the thing that makes it a little bit complicated is that the hard size and the polymorphic size are pushed onto the stack from different places (hard size is [in pxla.py](https://github.com/google/jax/blob/41f1eb412c2522be34795cd47148f3e39405088c/jax/interpreters/pxla.py#L699), polymorphic size is [in masking.py](https://github.com/google/jax/blob/c88e28c9a51fb1e2a3d12eb4f5e70164f6143a04/jax/interpreters/masking.py#L115)). I think the polymorphic size gets pushed first, so we'd have to have some logic to replace the tuple ('i', b ** 2 + a + 1) with ('i', 9, b ** 2 + a + 1) when the hard size gets pushed. I'm gonna have a go at implementing that just to see what it looks like, but it will be slightly more complicated than the current version.

Reply:
Going to resolve this but if you can see a better solution lmk and we can implement it!